File: 백준/Gold/9663. N－Queen/N－Queen.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

	static int n;
	static int result;
	static int[] column;

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		n = Integer.parseInt(br.readLine());
		column = new int[n];
		dfs(0);
		System.out.println(result);
	}

	static void dfs(int row) {

		if (row == n) {
			result++;
			return;
		}

		for (int i = 0; i < n; i++) {
			column[row] = i;
			if (isPossible(row)) {
				dfs(row + 1);
			}
		}
	}

	static boolean isPossible(int row) {
		for (int i = 0; i < row; i++) {
			if (column[i] == column[row] || Math.abs(column[row] - column[i]) == row - i) {
				return false;
			}
		}
		return true;
	}
}

File: 백준/Gold/9663. N－Queen/N－Queen.py
import sys

n = int(sys.stdin.readline().rstrip())
result = 0
column = [0 for _ in range(n)]


def isPossible(depth):
    for i in range(depth):
        if column[i] == column[depth] or abs(column[depth] - column[i]) == depth - i:
            return False
    return True


def dfs(depth):
    global result

    if depth == n:
        result += 1
        return

    for i in range(n):
        column[depth] = i
        if isPossible(depth):
            dfs(depth + 1)


dfs(0)
print(result)
